### PR TITLE
Add base64, base64url and hex encoding

### LINKED ISSUE #469: Add base64 and base64url and hex encoding & decoding
I believe the effect ecosystem would benefit from built-in base64 and base64url encoding & decoding utils.

### Status quo

The relationship status of the web platform and base64 can only be described as "it's complicated".

Most browsers natively provide two JavaScript functions for decoding and encoding Base64 *strings*(!): `btoa` (Binary to ASCII) and `atob` (ASCII to Binary). Notably, these [are both *not* part of the ECMAScript spec](https://html.spec.whatwg.org/multipage/webappapis.html#atob),

> From  https://developer.mozilla.org/en-US/docs/Glossary/Base64
> 
> [...] btoa and atob were added to the web platform before it supported binary data types. As a result, the two functions use strings to represent binary data, with the code point of each character representing the value of each byte. This has led to a common misconception that btoa can be used to encode arbitrary text data [...]
> 
> However, the byte-to-code-point correspondence only reliably holds true for code points up to 0x7f. Furthermore, code points over 0xff will cause btoa to throw an error due to exceeding the maximum value for 1 byte.

Both, `btoa` and `atob`, were also [added in Node.js v16 but immediately flagged as deprecated](https://nodejs.org/dist/latest-v16.x/docs/api/globals.html#atobdata) (Stability 3; Legacy). The Node.js documentation continues to recommend the good old `Buffer` APIs for base64 encoding & decoding.

There is currently a [Stage 2 ECMAScript Proposal](https://tc39.es/proposal-arraybuffer-base64/) to add base64 & hex encoding & decoding. 

### Recommended approach

I suggest we implement a base64, a base64url and a hex module and expose full encoding & decoding implementations without attempting to fall back to native `Buffer` or `atob` / `btoa` APIs. This guarantees cross platform functionality & reproducability until the TC39 base64 proposal is implemented. We can use the [Deno implementation](https://github.com/denoland/deno_std/blob/main/encoding/base64.ts) as a blueprint which itself is inspired by an [implementation by Egor Nepomnyaschih](https://gist.github.com/enepomnyaschih/54c437997f8202871278d0fdf68148ca).

I'm going to run some benchmarks to compare performance between `atob`, `btoa`, `Buffer` and the implementations mentioned above.

### Reference

- https://developer.mozilla.org/en-US/docs/Glossary/Base64
- https://html.spec.whatwg.org/multipage/webappapis.html#atob
- https://nodejs.org/dist/latest-v16.x/docs/api/globals.html#atobdata
- https://github.com/nodejs/node/blob/982e7a65cc68b45a0cfe45e2d7c895276501eba1/lib/buffer.js#L1259-L1354
- https://github.com/tc39/proposal-arraybuffer-base64
- https://deno.land/std@0.198.0/encoding

### PR DESCRIPTION
**Pull request by fubhy**
This implements:

- [x] base64 encoding (RFC4648) based on an [implementation by Egor Nepomnyaschih](https://gist.github.com/enepomnyaschih/54c437997f8202871278d0fdf68148ca) with some slight modifications & a minor bug fix.
- [x] base64url encoding (based on Deno)
- [x] hex encoding (based on Deno)

Fixes #469

### CODE REVIEW

File: docs/modules/Base64.ts.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+# encoding
+
+## decode
+
+Decodes a base64 (RFC4648) encoded string.
+
+**Signature**
+
+```ts
+export declare const decode: (str: string) => Either.Either<Error, Uint8Array>
```
Comment (fubhy) [inline docs/modules/Base64.ts.md:33]: Should this return a `Option` instead for simplicity? If it's idiomatically correct to return a `Either` here, would we want this to be a specific tagged error?

**Review by fubhy — COMMENTED** [review]
Questions:

- Is `Either` the right choice here idiomatically considering that the error case is usually rather obvious and would arguably be equally well described with `Option`?
- If we go with `Either` does this require a tagged error class?
- Do we want to add encoding & decoding interfaces & utils to deal with objects that implement these?

I'll wait for some initial feedback before I continue with the Base64Url scheme and Hex encoding.

**Comment by fubhy** [comment]
I think housing the Encoding module in `effect/data` would be preferable. This could  prove useful in internal modules that might not depend on `effect/schema` at that layer. E.g. for a hypotehtical future use case in the vicinity of the `effect/platform` http module ... Maybe for some kind of token encoding etc.? Only exposing it through `effect/schema` might not be desirable there.

File: src/Encoding.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * @category encoding
+ * @since 1.0.0
+ */
+export const encodeBase64 = Base64.encode
+
+/**
+ * @since 1.0.0
+ */
+export class Base64DecodeError {
+  /** @since 1.0.0 */
```
Comment (fubhy) [inline src/Encoding.ts:23]: Is this double tagging of the `@since` tag required or is that a bug? If I leave this out `docgen` fails.

**Comment by mikearnaldi** [comment]
> I think housing the Encoding module in `effect/data` would be preferable. This could prove useful in internal modules that might not depend on `effect/schema` at that layer. E.g. for a hypotehtical future use case in the vicinity of the `effect/platform` http module ... Maybe for some kind of token encoding etc.? Only exposing it through `effect/schema` might not be desirable there.
> 
> I re-organized it into `internal/Encoding/Base64.ts` now and exposed it publicly as `@effect/data/Encoding` per your suggestion 👍 .

Well `platform` etc will depend on schema at some point

**Comment by fubhy** [comment]
> Well `platform` etc will depend on schema at some point

Fair. Though my gut feeling is still that a standalone encode & decode utility may have its merits. At the end of the day it's easy to move these around. Your call ;-)

**Comment by tim-smart** [comment]
This might be better in the platform module, then you can use Effect<> for the api, and use the node:crypto api for a node implementation.

**Comment by fubhy** [comment]
> This might be better in the platform module, then you can use Effect<> for the api, and use the Buffer / node:crypto module for a node implementation.

Ok. Where would the browser implementation live? Just in @effect/platform? And the buffer one in @effect/platform-node?

**Comment by tim-smart** [comment]
Anything platform independent goes in /platform.

Eventually a /platform-browser will be added for browser only implementation.

**Comment by fubhy** [comment]
> Anything platform independent goes in /platform, as well as the model.
> 
> Eventually a /platform-browser will be added for browser only implementation.

Ok that makes sense! Will adjust and reopen in platform.

**Pull request closed without merging**